File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    void buildStringTree(TreeNode root, StringBuilder sb){
        if(root == null){
            sb.append("N ");
            return;
        }

        sb.append(Integer.toString(root.val) + " ");
        buildStringTree(root.left, sb);
        buildStringTree(root.right, sb);
    }

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder tree = new StringBuilder();
        buildStringTree(root, tree);
        return tree.toString();
    }

    TreeNode buildTree(int[] idx, String[] data){
        idx[0]++;

        if(idx[0] >= data.length || data[idx[0]].equals("N")){
            return null;
        }

        String currNodeVal = data[idx[0]];

        TreeNode root = new TreeNode(Integer.parseInt(currNodeVal));
        root.left = buildTree(idx, data);
        root.right = buildTree(idx, data);

        return root;
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] nodeValues = data.split(" ");
        int[] idx = new int[1];
        idx[0] = -1;
        return buildTree(idx, nodeValues);
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """

        result = []

        def pre_order(node):
            if node == None:
                result.append("N")
                return
            
            result.append(str(node.val))
            pre_order(node.left)
            pre_order(node.right)
            
        pre_order(root)
        return ",".join(result)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        self.i = -1
        data = data.split(",")

        def pre_order():
            self.i += 1
            if data[self.i] == "N":
                return None
            
            new_node = TreeNode(int(data[self.i]))
            new_node.left = pre_order()
            new_node.right = pre_order()
            return new_node

        return pre_order()



# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))